tiledb/sm/subarray/subarray.cc

```diff
@@ -1906,34 +1900,69 @@ Status Subarray::compute_tile_coords_row() {
   return Status::Ok();
 }
 
-Status Subarray::compute_tile_overlap(ThreadPool* const compute_tp) {
+Status Subarray::precompute_tile_overlap(
+    const uint64_t start_range_idx,
+    const uint64_t end_range_idx,
+    ThreadPool* const compute_tp,
+    const bool override_memory_constraint) {
   STATS_START_TIMER(stats::Stats::TimerType::READ_COMPUTE_TILE_OVERLAP)
 
-  const bool tile_overlap_computed = tile_overlap_ != nullptr;
-  if (tile_overlap_computed)
+  // If the `tile_overlap_` has already been precomputed and contains
+  // the given range, re-use it with new range.
+  const bool tile_overlap_computed =
+      tile_overlap_.contains_range(start_range_idx, end_range_idx);
+  if (tile_overlap_computed) {
+    tile_overlap_.update_range(start_range_idx, end_range_idx);
     return Status::Ok();
+  }
 
   compute_range_offsets();
 
-  // Initialization
-  auto tile_overlap = tdb_make_shared(std::vector<std::vector<TileOverlap>>);
   auto meta = array_->fragment_metadata();
   auto fragment_num = meta.size();
-  tile_overlap->resize(fragment_num);
-  auto range_num = this->range_num();
-  for (unsigned i = 0; i < fragment_num; ++i)
-    (*tile_overlap)[i].resize(range_num);
+
+  uint64_t tile_overlap_start = start_range_idx;
+  uint64_t tile_overlap_end = end_range_idx;
+
+  // Currently, we allow the caller to override the memory constraint
+  // imposed by `constants::max_tile_overlap_size`. This is temporary
+  // until we refactor for all callers to become aware that `tile_overlap_`
+  // may be truncated.
+  if (!override_memory_constraint && fragment_num > 0) {
+    // The size of each `TileOverlap` instance is variable. We estimate each
+    // instance size as if they have `1` tile and `1` range.
+    const uint64_t est_tile_overlap_size = TileOverlap::est_byte_size(1, 1);
```

Comment:
This is probably quite an underestimation, but probably it's ok. 

Another way to do it in the future is by computing tile overlaps gradually for the ranges, without estimation a priori. That is, compute `nthreads` number of range tile overlaps, get the sum of their sizes, check against a maximum parameter, and continue or break accordingly. Right now you are fixing the end range based on this estimation instead, which may be off.

Reply:
I'll ticket work to refactor as we discussed yesterday